chrome/content/zotero/xpcom/utilities.js

```diff
@@ -144,6 +144,238 @@ var CSL_TYPE_MAPPINGS = {
+	 * @param {Object} identifier
+	 * @return {Promise<Object|null>}
+	 */
+	"getItemFromAPI": async function (identifier) {
+		let translate = new Zotero.Translate.Search();
+		translate.setIdentifier(identifier);
+		
+		let translators = await translate.getTranslators();
+		translate.setTranslator(translators);
+		let newItems;
+		try {
+			newItems = await translate.translate({
+				libraryID: false
+			});
+		}
+		catch (e) {
+			Zotero.logError(e);
+			return null;
+		}
+		return newItems[0];
+	},
+	
+	/**
+	 * Update item metadata
+	 *
+	 * @param {Zotero.Item} item
+	 * @return {Promise<bool>}
+	 */
+	"updateMetadata": async function (item) {
+		// Get an identifier from the item
+		let identifier = Zotero.Utilities.getItemIdentifier(item);
+		if (!identifier) {
+			// If no identifier is found, use other item fields to resolve it
+			let resolvedIdentifier = await Zotero.Utilities.resolveIdentifier(item);
+			// For now, resolve only DOI identifiers
+			if (resolvedIdentifier.DOI) {
+				identifier = resolvedIdentifier;
+			}
+		}
+		
+		let newItem = null;
+		
+		if (identifier) {
+			let item1 = await Zotero.Utilities.getItemFromAPI(identifier);
+			let item2 = null;
+			
+			if (identifier.DOI) {
+				item2 = await Zotero.Utilities.getItemFromURL('https://dx.doi.org/' + identifier.DOI);
+			}
+			
+			// If both items are translated
+			if (item1 && item2) {
+				// Combine item1 and item2
+				for (let key in item2) {
+					if (Zotero.Utilities.fieldIsValidForType(key, item1.itemType)) {
+						// Add all item2 fields that doesn't exist in item1
+						if (!item1[key] && item2[key]) {
+							item1[key] = item2[key];
+						}
+					}
+				}
+				newItem = item1;
+			}
+			// If only item1 is translated
+			else if (item1) {
+				newItem = item1;
+			}
+			// If only item2 is translated
+			else if (item2) {
+				newItem = item2;
+			}
+		}
+		
+		// Translate by URL if translation by identifier failed
+		if (!newItem && item.getField('url')) {
+			newItem = await Zotero.Utilities.getItemFromURL(item.getField('url'));
+		}
+		
+		let updated = false;
+		if (newItem) {
+			let newItemTypeID = Zotero.ItemTypes.getID(newItem.itemType);
+			
+			if (item.itemTypeID !== newItemTypeID) {
+				item.setType(newItemTypeID);
+				updated = true;
+			}
+			
+			let creators1 = item.toJSON().creators;
+			let creators2 = newItem.creators;
+			
+			let creatorsUpdated = false;
+			
+			// Compare creators count
+			if (creators1.length !== creators2.length) {
+				creatorsUpdated = true;
+			}
+			else {
+				// Compare each creator
+				for (let i = 0; i < creators1.length; i++) {
+					if (
+						creators1[i].creatorType !== creators2[i].creatorType ||
+						creators1[i].firstName !== creators2[i].firstName ||
+						creators1[i].lastName !== creators2[i].lastName
+					) {
+						creatorsUpdated = true;
+						break;
+					}
+				}
+			}
+			
+			if (creatorsUpdated) {
+				// Empty creators, because setCreators() doesn't do that bey default
```

Comment:
That's probably a bug, though. I don't think there's a good reason for `setCreators()` to set the number of creators passed and leave the rest.

Reply:
Well, but the code seems intentionally made to clear creators only when passing empty array:
https://github.com/zotero/zotero/blob/master/chrome/content/zotero/xpcom/data/item.js#L1091

Reply:
Yeah, but I still think it's a bug. Calling it with 3 creators, you'd never expect creators 4 through 6 to remain on the item.